### PR TITLE
fix: guard empty batch (T=0) in Vulkan linear dispatch

### PR DESCRIPTION
**Pull request by maci0**
## Problem

`vulkan_ops.linear()` with an empty batch (`T == 0`) kills the process with
SIGSEGV. `T == 0` is `< _MATVEC_THRESHOLD`, so it dispatches `_vulkan_matvec`,
which computes `out_size = T * N * 4 == 0` and asks for a zero-byte Vulkan
output buffer. It is a hard crash, not a Python exception, so `_wrap_linear`'s
`try/except ... return orig(...)` cannot fall back to CPU.

vLLM's warmup/profiling passes call Linear with an empty batch, the same class
of degenerate call `_wrap_linear`'s `weight.numel() == 0` guard already handles
from the other side.

Measured on a 7900 XTX (RADV, gfx1100) at q_proj's shape (2048x1536), against
`torch.nn.functional.linear`:

```
T= 0  SIGSEGV
T= 1  max abs err 0.0299
T= 2  max abs err 0.0325
T= 3  max abs err 0.0327
T= 4  max abs err 0.0314
T= 8  max abs err 0.0333
T=17  max abs err 0.0352
```

Every `T >= 1` is correct (the ~0.03 error is the expected f16 weight-upload
rounding from #83). Only `T == 0` is broken.

## Fix

Return an empty result in `linear()` before the dispatch decision. CPU is
correct and free at this size; the `else` branch already handled `T == 0`
fine, it just wasn't reachable for a large weight.

## Tests

`TestLinearDispatchBoundary` in `tests/python/test_vulkan_ops.py` sweeps
`T = 0,1,2,3,4,8,17` with and without bias at 2048x1536 (above
`_MATVEC_MIN_WEIGHT_ELEMENTS`, so GPU dispatch is really exercised rather than
silently falling back to CPU). Without the guard, `T == 0` segfaults pytest.

## What changed from the previous version of this PR

The small-`M` padding in `model_runner.py` is gone. It was written against a
`linear()` that had no `T < 4` path; now that one exists, padding would push
decode *into* the tiled matmul rather than away from it. Rebased onto current
main.

### CODE REVIEW

**Comment by maci0** [comment]
Good catches, both fixed in the latest push:

- **Empty batch (M=0):** now returns an empty result before the padding path, so no more `IndexError` on `xf[-1:]` during warmup/empty iterations.
- **Weight caching:** switched from `weight.float()` to passing the original `weight`. Confirmed `vulkan_ops.linear` handles bf16/fp16/fp32 weights directly and caches by storage id, so the float cast was the thing forcing a re-upload every step.

Verified on a 7900 XTX (RADV): the original-dtype path matches `torch.nn.functional.linear` for M=1..8 across all three dtypes, and the test now covers M=0.

**Comment by ericcurtin** [comment]
Thanks for contributing @maci0 just poking you so you know this one needs work, didn't pass the build

**Comment by ericcurtin** [comment]
## Review

Root cause and fix look correct: `_vulkan_matvec`'s underlying shader tiles 4 rows, and padding+slicing in `_wrap_linear` is a reasonable dispatch-layer workaround. I checked the concerns that jumped out on first read and they turn out to be non-issues:

- Dropping the explicit `weight.float()` / relying on `vulkan_ops.linear()` to do the float conversion internally is fine — `linear()` (vulkan_ops.py:239,248,264) always calls `.float()` on `x`/`weight`/`bias` itself, so nothing is lost.
- `xf[-1:].expand(4 - m, -1)` produces a stride-0 view, but `_to_bytes()` calls `.contiguous()` before serializing, so that's safe too.

### Blocking
CI is red: `ruff` `I001` (unsorted import block) in `tests/python/test_linear_dispatch.py:22-23` on all 3 lint jobs. Trivial — `ruff check --fix` or reorder the `vulkan_ops`/`VulkanContext` imports.

### Needs coordination with #19
This PR and #19 both rewrite the same lines of `vk_forward` in `vllm_vulkan/model_runner.py` (the `bias = getattr(...)` / try-block / tuple-return section). I test-merged both locally and confirmed a real conflict:

```
git merge --no-commit --no-ff pr19   # on top of pr17
CONFLICT (content): Merge conflict in vllm_vulkan/model_runner.py
```

Neither PR is wrong on its own, but whichever merges second will need a manual rebase that keeps *both* the small-`M` padding and the TP collective logic together (i.e. pad-then-matmul-then-collective, not one replacing the other). Worth flagging to the author/reviewer merging these so the fix isn't silently dropped.

**Mergeability: 8/10** — correct fix, tests included; blocked only by a one-line lint fix and coordinating the rebase with #19.

**Comment by ericcurtin** [comment]
## Review (re-check against current main)

### Conflicts
`vllm_vulkan/model_runner.py` conflicts with main, where `vk_forward` was rewritten (TP collectives, tuple return, weight-cache handling). Needs a rebase before anything else.

### The fix is now obsolete
Main's `vulkan_ops.linear` (`vulkan_ops.py:644-655`) already dispatches on `T`:
- `T < _MATVEC_THRESHOLD` (= 4) goes to `mul_mat_vec_{f16,f32}_f32_f32`
- `T >= 4` goes to the tiled `mul_mm` matmul

So the small-`M` tiled-matmul path this pads around is no longer reachable for decode. Padding `M` up to 4 would actively push decode *into* the tiled matmul: slower, and the exact path you measured as wrong.

The `weight.float()` to `weight` change also already landed on main (`model_runner.py:376-396`, with the cache-identity rationale documented).

### Also
`ruff I001` in `tests/python/test_linear_dispatch.py:22-23`, so lint is red on all three jobs.

### Suggestion
Rebase and reduce this to just `tests/python/test_linear_dispatch.py` as a regression guard over the `M = 0,1,2,3,4,8,17` dispatch boundary (it's genuinely useful, nothing on main covers `M` around the threshold), dropping the padding in `model_runner.py`. If the M<4 garbage still reproduces on current main, please reopen with that measurement instead; otherwise this can be closed.

**Comment by ericcurtin** [comment]
Re-checked against current main; still blocked:

- Conflicts in `vllm_vulkan/model_runner.py` — needs a rebase.
- Lint red: `ruff I001` in `tests/python/test_linear_dispatch.py:22-23`.
- The padding is now counterproductive: `vulkan_ops.linear` already sends `T < 4` to the matvec shader, so padding to 4 forces decode into the tiled matmul. Please drop the `model_runner.py` change and keep only the `M = 0..17` regression test.

Moving to draft; please mark ready for review again once rebased and green.

**Comment by maci0** [comment]
Rebased onto main and reworked along the lines you asked for, with one addition.

You were right that the padding is obsolete: I re-measured on a 7900 XTX (RADV, gfx1100) at q_proj's shape and every `T >= 1` matches `torch.nn.functional.linear` on current main (max abs error ~0.03, which is just the f16 weight-upload rounding from #83). So the `model_runner.py` change is dropped.

What the boundary sweep did turn up is `T == 0`: it's `< _MATVEC_THRESHOLD`, so it dispatches `_vulkan_matvec`, which asks for a zero-byte output buffer (`out_size = T * N * 4`) and takes the process down with SIGSEGV. Not a catchable exception, so `_wrap_linear`'s try/except can't fall back to CPU. vLLM's warmup/profiling passes do hit Linear with an empty batch. The fix is a `T == 0` early return in `linear()` before the dispatch decision.

The test is `TestLinearDispatchBoundary` in `tests/python/test_vulkan_ops.py` rather than a new `test_linear_dispatch.py`: it reuses the existing `vulkan_ctx` fixture and sits next to `TestLinearMatvecDispatchThreshold` / `TestLinearTiledMatmulPrefillDispatch`, which is where someone changing the threshold will look. Happy to move it to its own file if you'd rather.

Local state: `ruff check`, `ruff format --check` and `mypy` clean. `pytest -m "not slow" tests/python/` goes from 96 passed / 12 failed on clean main to 110 passed / 12 failed. The 12 are pre-existing `TestFusedSwigluMlp` failures on this hardware, unrelated to this change; I can open a separate issue with the details if that's useful.

**Review by ericcurtin — APPROVED** [review]
Reproduced T=0 failing on main and passing with this branch on NVIDIA GB10; full sweep 14/14. Merging.

Follow-up: `_fused_swiglu_mlp` (`vulkan_ops.py:989`) has the same T=0 hazard (`gate_up_out_size = T * ... = 0`). Same early return, or raise `FusedMlpUnavailableError` so `_wrap_qwen_mlp` falls back.
